### DISCUSSION TITLE
Strange behavior from Serial

### DISCUSSION THREAD

**Discussion by Sandvoxel** [SpenceKonde/megaTinyCore#1069] [Q&A]
I will start out by saying I have a bit of a strange setup where my serial lines are connected to a FT260-Q ftdi chip that I am manually reading the interrupts from to grab my data but i do seem to be able to read serial just find with a script in the Arduino editor(I am currently using platformio for the main project). This FT260-Q is always connected even during programing as long as I dont close the program I wrote to read the data.(I am using pymcuprog uart programmer but this is a completely separate ftdi uart bridge). I am experiencing a strange issue where the the data I get back is almost good but will get random hiccups.

![image](https://github.com/SpenceKonde/megaTinyCore/assets/15006020/cc0f7da1-9a48-4cd8-840a-e444f72228b8)

The strange part to me is if I open my program that read in the serial as the chip is programing it will read out the data correctly.

![image](https://github.com/SpenceKonde/megaTinyCore/assets/15006020/7c432a69-95fe-4a47-9a11-07ede99ea88d)

At first I thought it was a buffer overflow but the packet size is limited to a max of 32 bytes

Here is the code that transmits the data to serial for completion 
```cpp
void send(const uint8_t* buffer, size_t size) const
{
    if(_stream == nullptr || buffer == nullptr || size == 0) return;

    uint8_t _encodeBuffer[EncoderType::getEncodedBufferSize(size)];

    size_t numEncoded = EncoderType::encode(buffer,
                                            size,
                                            _encodeBuffer);

    _stream->write(_encodeBuffer, numEncoded);
    _stream->write(PacketMarker);
}


void sendPacket(const uint8_t* data, size_t length, uint8_t packetID) {
    const size_t maxPacketSize = 32; // Total maximum packet size
    const size_t maxPayloadSize = maxPacketSize - 4; // Adjusting for 2 bytes for packet ID/index and 2 bytes for CRC

    // Safeguard: Limit the total data length to a maximum (optional)
    const size_t maxDataLength = 1024; // Example maximum data length
    length = min(length, maxDataLength);

    size_t dataSent = 0; // Keep track of how much data has been sent
    uint8_t status;

    while (dataSent < length) {
        // Calculate remaining data size
        size_t remainingData = length - dataSent;
        size_t chunkSize = (remainingData > maxPayloadSize) ? maxPayloadSize : remainingData;
        status = (chunkSize == remainingData) ? 0xFF : 0;

        // Prepare packet buffer, ensuring it does not exceed maxPacketSize
        uint8_t packet[maxPacketSize]; // Buffer for packet ID + packet index + payload + CRC

        // Prepend packet ID and packet index
        packet[0] = packetID;
        packet[1] = status;

        // Copy chunk of data into the packet
        memcpy(packet + 2, data + dataSent, chunkSize);

        // Calculate CRC for this chunk + packet ID and index
        uint16_t crc = crc16(packet, chunkSize + 2); // Include packet ID and index in CRC calculation

        // Append CRC to the end of the packet
        packet[chunkSize + 2] = crc & 0xFF;
        packet[chunkSize + 3] = (crc >> 8) & 0xFF;

        // Send the packet
        myPacketSerial.send(packet, chunkSize + 4);

        // Update counters
        dataSent += chunkSize;

    }
}
```

I quite at a loss here on why the case where If open my software while it is programing causes it to work better. I would think that it has something to do with the uart alignment because my program re-sets up the uart chip but I dont see how that would be. Maybe I am missing something here.

```ini
platform = https://github.com/platformio/platform-atmelmegaavr.git
framework = arduino
board = ATtiny3227
board_build.mcu = attiny3227
board_build.f_cpu = 10000000L
upload_speed = 230400
build_flags =
;	-D RELEASE
upload_flags =
	--tool
	uart
	--device
	attiny3227
	--uart
	$UPLOAD_PORT
	--clk
	$UPLOAD_SPEED
upload_command = pymcuprog write --erase $UPLOAD_FLAGS --filename $SOURCE
board_hardware.bod = 2.7v
board_hardware.eesave = yes
```

**Comment by Sandvoxel** [comment]
It also seems that sometime a packet will just not be sent at all but the following packet from the same call will.

**Comment by pcfreak1201** [comment]
Hi Sandvoxel!
Just for information: you use PlatformIO, not Arduino1.8.x - so it is not supported by Spence directly
On the other side don't you select the megatinycore at all, when I read this correctly:
https://registry.platformio.org/tools/platformio/framework-arduino-megaavr-megatinycore/installation
May be there is the problem? Just my two cents ...
